lib/ExtraCorres.thy

```diff
@@ -203,4 +204,40 @@ lemma wp_from_corres_unit:
   f' \<lbrace>\<lambda>_ s'. \<exists>s. (s,s') \<in> state_relation \<and> Q s \<and> Q' s'\<rbrace>"
   by (auto intro!: wp_from_corres_u_unit)
 
+definition
+  "ex_abs_underlying sr P s' \<equiv> \<exists>s. (s,s') \<in> sr \<and> P s"
+
+lemma ex_absI[intro!]:
+  "(s, s') \<in> sr \<Longrightarrow> P s \<Longrightarrow> ex_abs_underlying sr P s'"
+  by (auto simp add: ex_abs_underlying_def)
+
+lemma corres_underlying_split_ex_abs:
+  assumes ac: "corres_underlying srel nf nf' r' G G' a c"
+  assumes valid: "\<lbrace>G\<rbrace> a \<lbrace>P\<rbrace>" "\<lbrace>G' and ex_abs_underlying srel G\<rbrace> c \<lbrace>P'\<rbrace>"
+  assumes bd: "\<forall>rv rv'. r' rv rv' \<longrightarrow>
+                        corres_underlying srel nf nf' r (P rv) (P' rv') (b rv) (d rv')"
+  shows "corres_underlying srel nf nf' r G G' (a >>= (\<lambda>rv. b rv)) (c >>= (\<lambda>rv'. d rv'))"
+  using ac bd valid
+  apply (clarsimp simp: corres_underlying_def bind_def)
+  apply (clarsimp simp: Bex_def Ball_def valid_def ex_abs_underlying_def)
+  by meson
+
+lemma hoare_from_abs:
+  assumes abs_valid: "\<lbrace>P and R\<rbrace> f \<lbrace>\<lambda>rv. Q rv and S\<rbrace>"
+  assumes cross: "\<And>s s'. \<lbrakk>(s, s') \<in> srel; P' s'; R' s'\<rbrakk> \<Longrightarrow> P s"
+                 "\<And>s s' r r'. \<lbrakk>(s, s') \<in> srel; rrel r r'; Q r s; S s\<rbrakk> \<Longrightarrow> Q' r' s'"
+  assumes corres: "corres_underlying srel False True rrel G G' f f'"
+  shows "\<lbrace>P' and G' and R' and ex_abs_underlying srel (G and R)\<rbrace> f' \<lbrace>Q'\<rbrace>"
+  apply (insert assms)
+  by (fastforce simp: valid_def ex_abs_underlying_def corres_underlying_def)
+
+lemma hoare_from_abs_inv:
+  assumes abs_valid: "f \<lbrace>P\<rbrace>"
+  assumes cross: "\<And>s s'. (s, s') \<in> srel \<Longrightarrow> P s = P' s'"
+  assumes corres: "corres_underlying srel False True rrel G G' f f'"
+  shows "\<lbrace>P' and G' and ex_abs_underlying srel G\<rbrace> f' \<lbrace>\<lambda>_. P'\<rbrace>"
+  by (insert assms)
```

Comment:
this is a bit of a strange formulation, probably this would be more "canonical":
```
using assms
by (rule ... ,
    fastforce+)  (or ; fastforce if the reverse-order doesn't mess it up)
```

Reply:
This didn't work, unfortunately, and I can't find anything much nicer than what was there originally

Reply:
After rearranging some of the assumptions in other rules and using some extra manual instantiation, I got this working